### PR TITLE
fixed autoscaling bug and updated hpa sample

### PR DESCRIPTION
**Pull request by asteurer**

### CODE REVIEW

**Comment by asteurer** [comment]
@bacongobbler @calebschoepp I removed the requirements for a minimum pod count, but I'm running into an issue with `spec.replicas`. When I don't set the replica field, all resources are created; however, I can't port-forward the service, which is how I'm testing the scaling capabilities. Any thoughts?

**Review by michelleN — COMMENTED** [review]
@asteurer is there an issue I can look at to understand the bug?

**Comment by michelleN** [comment]
>@asteurer is there an issue or thread I can look at to understand the bug?

nvm found #24 !

File: config/samples/hpa.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ metadata:
 spec:
   image: ghcr.io/spinkube/spin-operator/cpu-load-gen:20240311-163328-g1121986
   executor: containerd-shim-spin
-  enableAutoscaling: true
+  replicas: 1
```
Comment (calebschoepp) [inline config/samples/hpa.yaml:8]: @bacongobbler says:
> Shouldn't this be removed if both `minReplicas` and `maxReplicas` are set?

File: internal/controller/spinapp_controller.go [hunk 2]
```diff
@@ -381,14 +391,7 @@ func (r *SpinAppReconciler) deleteDeployment(ctx context.Context, app *spinv1alp
...
 	generatedRuntimeConfigSecretName, caSecretName string, scheme *runtime.Scheme) (*appsv1.Deployment, error) {
-	// TODO: Once we land admission webhooks write some validation to make
-	// replicas and enableAutoscaling mutually exclusive.
-	var replicas *int32
-	if app.Spec.EnableAutoscaling {
-		replicas = nil
-	} else {
-		replicas = ptr(app.Spec.Replicas)
-	}
+	replicas := app.Spec.Replicas
```
Comment (calebschoepp) [inline internal/controller/spinapp_controller.go:394]: @bacongobbler says:
> related to the above comment RE: min/maxReplicas, shouldn't this be nil if either of those are set?
> 
> If only minReplicas is set (is that valid?), should this be set to minReplicas as a starting value?
> If only maxReplicas is set, should this be set to 1 (or possibly nil, assuming we can scale to zero)?

File: api/v1alpha1/spinapp_types.go [hunk 3]
```diff
@@ -99,6 +91,9 @@ type SpinAppStatus struct {
 
 	// Represents the current number of active replicas on the application deployment.
 	ReadyReplicas int32 `json:"readyReplicas"`
+
+	// Selector for the underlying pods
+	Selector string `json:"selector"`
```
Comment (calebschoepp) [inline api/v1alpha1/spinapp_types.go:96]: @endocrimes would we have to make this optional to support runtimes where this selector is meaningless? Or is setting it to the empty string sufficient.

**Comment by calebschoepp** [comment]
> @bacongobbler @calebschoepp I removed the requirements for a minimum pod count, but I'm running into an issue with `spec.replicas`. When I don't set the replica field, all resources are created; however, I can't port-forward the service, which is how I'm testing the scaling capabilities. Any thoughts?

Maybe you can't port forward the service because the replicas defaults to zero, so there are no pods, so there is nothing to port forward. It's a bit circular haha.

You could try playing with KEDA scaling on some external factor like messages in a queue. That way to generate load you just put a bunch of messages in a queue and then it would scale it up from zero.

**Comment by asteurer** [comment]
> > Maybe you can't port forward the service because the replicas defaults to zero, so there are no pods, so there is nothing to port forward. It's a bit circular haha.
>

Shouldn't there be at least 1 pod since we set a minReplicas? If so, I'm not clear on how we would reference the HPA in the SpinApp deployment haha

**Comment by calebschoepp** [comment]
> Shouldn't there be at least 1 pod since we set a minReplicas? If so, I'm not clear on how we would reference the HPA in the SpinApp deployment.
> 
> Also, `k scale spinapp` does work with the updated minimum pod count logic

[This](https://stackoverflow.com/questions/61596711/in-kubernetes-how-can-i-scale-a-deployment-to-zero-when-idle) looks relevant.

File: config/samples/hpa.yaml (hunk 1 above)
Comment (calebschoepp) [reply config/samples/hpa.yaml:8]: @asteurer ?

File: internal/controller/spinapp_controller.go (hunk 2 above)
Comment (calebschoepp) [reply internal/controller/spinapp_controller.go:394]: @asteurer ?

**Review by calebschoepp — CHANGES_REQUESTED** [review]
No new comments from me. I just want you to address @bacongobbler's comments

**Comment by asteurer** [comment]
There's a bug in SpinKube that is preventing metrics from being properly emitted from Spin applications, which is preventing autoscaling, and is making it impossible for me to test. Kate Goldenring is working with jsturtevant (slack handle) from MS to get this fixed.

**Pull request closed without merging**